Doc/refresh debugging guidelines (#6082)
solo-io/gloo: docs/content/operations/debugging_gloo/_index.md

weight: 10
---

At times, you may need to debug Gloo Edge misconfigurations. Gloo Edge is based on [Envoy](https://www.envoyproxy.io) and often times these misconfigurations are observed as a result of behavior seen at the proxy. This guide will help you debug issues with Gloo Edge and Envoy. 

The guide is broken into three main sections:

* General debugging tools and tips
* Debugging the data plane 
* Debugging the control plane

This guide is intended to help you understand where to look if things aren't working as expected. After going through, if [all else fails](#all-else-fails), you can capture the state of Gloo Edge configurations and logs and join us on our [Slack](https://slack.solo.io) and one of our engineers will be able to help.

## General debugging tools and tips

> If you're experiencing unexpected behavior after installing and configuring Gloo Edge, the first thing to do is verify [installation]({{< versioned_link_path fromRoot="/installation/" >}}) and configuration. The fastest way to do that is to run the `glooctl check` [command]({{< versioned_link_path fromRoot="/reference/cli/glooctl_check/" >}}). This command will go through the deployments, pods and Gloo Edge resources to make sure they're in a healthy/Accepted/OK status. Typically if there are some problem syncing resources, you would find an issue here.

Reviewer: ```suggestion
If you're experiencing unexpected behavior after installing and configuring Gloo Edge, the first thing to do is verify [installation]({{< versioned_link_path fromRoot="/installation/" >}}) and configuration. The fastest way to do that is to run the `glooctl check` [command]({{< versioned_link_path fromRoot="/reference/cli/glooctl_check/" >}}). This command will go through the deployments, pods and Gloo Edge resources to make sure they're in a healthy/Accepted/OK status. Typically if there are problems syncing resources, you would find an issue here.
```

glooctl proxy dump
```

This dumps the entire Envoy configuration, including all static and dynamic resources. Typically at the bottom you can see the VirtualHost and Route sections to verify your settings were picked up correctly.

> A more advanced way of generating the Envoy config dump is to port-forward to one of the gateway-proxy pods and to run the following command(s):

Reviewer: ```suggestion
A more advanced way of generating the Envoy config dump is to port-forward to one of the gateway-proxy pods and to run the following commands:
```

```bash
# 1. pick a gateway-proxy pod
kubectl -n gloo-system get pod -l "gloo=gateway-proxy"
# 2. port-forward on port 19000
kubectl -n gloo-system port-forward <pod name> 19000 &

> # 3. generate the config dump

Reviewer: ```suggestion
# 3a. generate the config dump
```

curl -X POST 127.0.0.1:19000/config_dump > gateway-config.json

> # 3bis. optionally include the upstream endpoints in the config dump

Reviewer: ```suggestion
# 3b. optionally include the upstream endpoints in the config dump
```

glooctl proxy logs -f
```

When you have the logging window up, send requests through to the proxy and you can get some very detailed debugging logging going through the log tail.

{{% notice warning %}}

> Keep in mind that this command will actually [change](https://github.com/solo-io/gloo/blob/c2e025728df3c66c67275ac718e251a275d32bd3/projects/gloo/cli/pkg/cmd/gateway/logs.go#L65) the log level to `debug`. You may want to revert it to `info` after that (see below).

Reviewer: ```suggestion
Keep in mind that this command will actually [change](https://github.com/solo-io/gloo/blob/c2e025728df3c66c67275ac718e251a275d32bd3/projects/gloo/cli/pkg/cmd/gateway/logs.go#L65) the log level to `debug`. You might want to revert it to `info` after that, as shown in the following commands.
```

{{% /notice %}}

A more advanced way of changing the log level, globally or on a per-logger basis, is through the Envoy Admin endpoints:

```bash
# 1. port-forward to the gateway-proxy pod
kubectl -n gloo-system port-forward deploy/gateway-proxy 19000 &
# 2. globally change the log level to debug
curl -X POST "127.0.0.1:19000/logging?level=debug"
# 3. change the log level only for the aws logger
curl -X POST "127.0.0.1:19000/logging?aws=debug"
```

For a full list of the different Envoy loggers, visit the following endpoint: `http://localhost:19000/logging`

> Additionally, you can configure access logging to dump specific parts of the request into the logs. Please see the [doc on access logging]({{< versioned_link_path fromRoot="/guides/security/access_logging//" >}}) to configure that. 

Reviewer: ```suggestion
Additionally, you can configure access logging to dump specific parts of the request into the logs. For more information, see [Access Logging]({{< versioned_link_path fromRoot="/guides/security/access_logging//" >}}). 
```

discovery-68dbd794-ssx7b                               1/1     Running   0          107m
extauth-67557744dd-5wc2p                               1/1     Running   0          107m
gateway-595cc67f54-tr6ps                               1/1     Running   0          107m
gateway-proxy-79c9f44b5d-cprg7                         1/1     Running   0          107m
gloo-74bb8b9df7-72t8m                                  1/1     Running   0          107m
gloo-fed-857964dd9f-gq8np                              1/1     Running   0          107m
gloo-fed-console-6f99dddccd-ls64k                      3/3     Running   0          107m
glooe-grafana-865bb9cd45-cdshq                         1/1     Running   0          107m
glooe-prometheus-kube-state-metrics-55ffc89cbb-kr8jx   1/1     Running   0          107m
glooe-prometheus-server-7d5b85764c-2nl2w               2/2     Running   0          107m
observability-5f8ffc8bdc-zggxb                         1/1     Running   0          107m
rate-limit-6d66688567-5tcx8                            1/1     Running   3          107m
redis-57fd559c5c-hcd6n                                 1/1     Running   0          107m
```

> Each component logs about the sync loops it runs (syncing with various environment signals like the Kube API, or Consul, etc). 

Reviewer: ```suggestion
Each component logs the sync loops that it runs, such as syncing with various environment signals like the Kubernetes API, Consul, etc. 
```

glooctl debug logs -f gloo-logs.log
glooctl debug yaml -f gloo-yamls.yaml
```

> This will dump all of the relevant configuration into to files, `gloo-logs.log` and `gloo-yamls.yaml` which gives a complete picture of your deployment. 

Reviewer: ```suggestion
These commands dump all of the relevant configuration into `gloo-logs.log` and `gloo-yamls.yaml` files, which gives a complete picture of your Gloo Edge deployment. 
```
